onnxscript/ir/passes/_pass_infra.py

```diff
@@ -97,76 +158,105 @@ def ensures(self, model: ir.Model) -> None:
         del model  # Unused
 
 
-class PassManager:
+class InPlacePass(PassBase):
+    """A pass that modifies the input model in place and returns it."""
+
+    @property
+    def in_place(self) -> bool:
+        return True
+
+    @property
+    def changes_input(self) -> bool:
+        return True
+
+
+class FunctionalPass(PassBase):
+    """A pass that returns a new model but does not modify the input model."""
+
+    @property
+    def in_place(self) -> bool:
+        return False
+
+    @property
+    def changes_input(self) -> bool:
+        return False
+
+
+class Sequential(PassBase):
+    """Run a sequence of passes in order."""
+
+    def __init__(self, *passes: PassBase):
+        if not passes:
+            raise ValueError("Sequential must take at least one pass")
+        self.passes = passes
+        self._in_place = all(pass_.in_place for pass_ in passes)
+        # The reason changes_inputs is decided by the first pass is that if the first pass is either in-place,
+        # or if it is not designed to be in-place but somehow changes the input (destructive),
+        # this pass sequence will change inputs.
```

Comment:
I mean would second or other passes that changes inputs after the first pass?

Reply:
If the first pass is functional, the second pass will take the new model that the first pass returns, which means the second pass has no chance to affect the input model.

Reply:
But if the first pass is side-effect only pass. Shouldn't we check the following passes?

Reply:
Good point. In that case we just assume it changes the model for now. I think that's ok?